libcxx/src/support/runtime/exception_pointer_cxxabi.ipp

```diff
@@ -34,6 +56,37 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
     return *this;
 }
 
+#  if defined(_LIBCPP_EXCEPTION_PTR_DIRECT_INIT)
+void *exception_ptr::__init_native_exception(size_t size, type_info* tinfo, void (_LIBCXX_DTOR_FUNC* dest)(void*)) noexcept
+{
+    __cxa_exception *(*cxa_init_primary_exception_fn)(void*, std::type_info*, void(_LIBCXX_DTOR_FUNC*)(void*)) = __cxa_init_primary_exception;
```

Comment:
We have a usual way of dealing with newly-introduced symbols. I already explained it in my other comment about the `__availability` header. Please use that instead. There is no need for us to check whether the weak symbol exists here and do something different at runtime in case it's not. The only conditional bit we should have is in `exception_ptr.h` where we would do

```
#if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
// use it
#else
// do the throw dance
#endif
```

Reply:
Do i understand you correctly that for libcxxrt/libsupc++ we would still need some kind of a runtime lookup?

I'm also confused about `LD_PRELOAD` case: is it okay that ld_preloading cxxabi of some previous version won't work?

Reply:
Both libcxxrt and libsupc++ do implement the function, so this should not be an issue, right?

If you `LD_PRELOAD` an older version of libc++abi, then you must have provided the appropriate deployment target when compiling your program. If you do and the deployment target doesn't provide the new function, there's no issue cause the code will use the throw-dance. If you lie about the deployment target (e.g. use a recent one) but then run against an old libc++abi, you get a load-time error for the missing symbol, as you should.

Reply:
And i've adjusted that to account recent changes in `__availability`.